src/openms/source/ANALYSIS/MAPMATCHING/QTClusterFinder.cpp

```diff
@@ -294,50 +301,72 @@ namespace OpenMS
-         it != clustering.end(); ++it)
+
+    const QTCluster& best = cluster_heads.top();
+
+    ClusterElementsMap const elements = best.getElements();
+
+    #ifdef DEBUG_QTCLUSTERFINDER
+    std::cout << "Elements: " << elements.size() << " with best "
+         << best->getQuality() << " invalid " << best->isInvalid() << std::endl;
+    #endif
+
+    createConsensusFeature_(feature, best.getCurrentQuality(), elements);
+
+    #ifdef DEBUG_QTCLUSTERFINDER
+    std::cout << " create new consensus feature " << feature.getRT() << " " << feature.getMZ() << " from " << best->getCenterPoint()->getFeature().getUniqueId() << std::endl;
+    for (OpenMSBoost::unordered_map<Size, OpenMS::GridFeature*>::const_iterator
+         it = elements.begin(); it != elements.end(); ++it)
     {
-      if (!it->isInvalid())
-      {
-        if (it->getQuality() > best->getQuality())
-        {
-          best = it;
-        }
-      }
+      std::cout << "   = element id : " << it->second->getFeature().getUniqueId() << std::endl;
     }
+    #endif
+
+    updateClustering_(element_mapping, grid, elements, cluster_heads, handles, best.getId());
 
-    // no more clusters to process -> clear clustering and return
-    if (best == clustering.end())
+    return true;
+  }
+
+  void QTClusterFinder::removeTopFromHeap_(Heap& cluster_heads,
+                                           const QTCluster& cluster,
+                                           ElementMapping& element_mapping)
+  {
+    ClusterElementsMap const elements = cluster.getElements();
+    for (ClusterElementsMap::const_iterator feature_it = elements.begin(); 
+        feature_it != elements.end(); ++feature_it)
     {
-      clustering.clear();
-      return;
+      unordered_set<Size>& cluster_ids = element_mapping[feature_it->second.second];
+      cluster_ids.erase(cluster.getId());
     }
 
-    OpenMSBoost::unordered_map<Size, OpenMS::GridFeature*> elements;
-    best->getElements(elements);
-#ifdef DEBUG_QTCLUSTERFINDER
-    std::cout << "Elements: " << elements.size() << " with best "
-         << best->getQuality() << " invalid " << best->isInvalid() << std::endl;
-#endif
+    cluster_heads.pop();
+  }
 
-    // create consensus feature from best cluster:
-    feature.setQuality(best->getQuality());
-    for (OpenMSBoost::unordered_map<Size, OpenMS::GridFeature*>::const_iterator
+void QTClusterFinder::createConsensusFeature_(ConsensusFeature& feature, double quality, const ClusterElementsMap& elements)
+  {
+    feature.setQuality(quality);
+    for (ClusterElementsMap::const_iterator
          it = elements.begin(); it != elements.end(); ++it)
     {
-      BaseFeature& elem_feat = const_cast<BaseFeature&>(it->second->getFeature());
+      // Store the id of already used features (important: needs to be done
+      // before updateClustering())
+      already_used_.insert(it->second.second);
+
+      BaseFeature& elem_feat = const_cast<BaseFeature&>(it->second.second->getFeature());
```

Comment:
this const_cast looks very dodgy....
what is the problem?

Reply:
I agree, didn't touch it, don't know what it does.

Reply:
I prefer not to change this, because I might have to change the interface of the GridFeature Class. This could potentially lead to a lot of changes elsewere and even more updated tests...